library/pk_wrap.c

```diff
@@ -526,13 +526,14 @@ static size_t eckey_get_bitlen(mbedtls_pk_context *pk)
 /* Common helper for ECDSA verify using PSA functions. */
 static int ecdsa_verify_psa(unsigned char *key, size_t key_len,
                             psa_ecc_family_t curve, size_t curve_bits,
+                            mbedtls_md_type_t md_alg,
                             const unsigned char *hash, size_t hash_len,
                             const unsigned char *sig, size_t sig_len)
 {
     int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
     psa_key_attributes_t attributes = PSA_KEY_ATTRIBUTES_INIT;
     mbedtls_svc_key_id_t key_id = MBEDTLS_SVC_KEY_ID_INIT;
-    psa_algorithm_t psa_sig_md = PSA_ALG_ECDSA_ANY;
+    psa_algorithm_t psa_sig_md = PSA_ALG_ECDSA(mbedtls_md_psa_alg_from_type(md_alg));
```

Comment:
I don't understand the point of this change. When signing/verifying an already-calculated hash, `PSA_ALG_ECDSA_ANY` and `PSA_ALG_ECDSA(hash_alg)` are equivalent.

The only difference I can think of is that `PSA_ALG_ECDSA(hash_alg)` enforces that the input has the expected length, but `mbedtls_pk_verify` does this anyway.

Also, `md_alg` can be `MBEDTLS_MD_NONE` at this point. In this case `psa_sig_md = PSA_ALG_ECDSA(0)` which happens to have the same value as `PSA_ALG_ECDSA_ANY` so it works, but it's relying on a numerical coincidence.

Reply:
The issue is that the cryptocell driver is more strict on the size, or basically doesn't support ECDSA_ANY, due to this check: 
https://git.trustedfirmware.org/TF-M/trusted-firmware-m.git/tree/lib/ext/cryptocell-312-runtime/codesafe/src/psa_driver_api/src/cc3xx_psa_asymmetric_signature.c#n462
@adeaarm Maybe that should be fixes the cryptocell driver as well

Reply:
Ok, if you want the code to work with a PSA implementation that supports `PSA_ALG_ECDSA(hash_alg)` but not `PSA_ALG_ECDSA_ANY`, that patch does help. But it only works if the caller of `mbedtls_pk_verify` passes an explicit `md_alg`, so there would be an undocumented limitation. It would be much better for Cryptocell to support `PSA_ALG_ECDSA_ANY`. We don't have precise rules for what mechanisms a PSA implementation should support, but having limitations like “`PSA_ALG_ECDSA(hash_alg)` but not `PSA_ALG_ECDSA_ANY`” is hard to work with.

We might be willing to accept this patch in Mbed TLS, but:

* We would need a comment explaining why we're doing something strange.
* I guess we would want to do signature as well.
* We have no way to test it, so it would be solely for the sake of one project, with the understanding that if it breaks in the future it's your responsibility to contribute a fix.
* This increases the complexity and code size, even if it's by a little bit, and I'm not yet convinced it's worth it.

Reply:
If I understand correctly the issue, the check that we do in CryptoCell is the one that Gilles mentions above explicitly:

> The only difference I can think of is that PSA_ALG_ECDSA(hash_alg) enforces that the input has the expected length

i.e. `IF hash_length != PSA_HASH_LENGTH(PSA_ALG_SIGN_GET_HASH(alg) THEN <INVALID_ARGUMENT>`

If this is too restrictive and does not bring any additional gain (i.e. an invalid length will anyway fail on the ECDSA verification), I think we can just relax the CC3XX check. Please feel free to open a patch there and I will review and merge it.

Reply:
This check does add a bit of protection against mistakes in application code, and we have it in Mbed TLS. What's apparently missing in CC is handling of `PSA_ALG_ECDSA_ANY` (which is numerically equal to `PSA_ALG_ECDSA(0)`). For that value, the input can be any length.

For security, the input to `PSA_ALG_ECDSA_ANY` must be a hash (ECDSA is insecure if applied directly on a malleable message), but for all you know, it's a hash function that the caller implemented and that doesn't exist in PSA, so you have no idea about its length.

Reply:
Thanks - having reviewed again the checks we do in that `CC3XX` function i believe they require some reorganization anyway (i.e. the check for `PSA_ALG_IS_ECDSA(alg))` happens not at the right place, so I think the most appropriate way for fixing this *is* to fix the checks as above: `PSA_ALG_IS_ECDSA(alg)` must be done earlier, and `alg == PSA_ALG_ECDSA_ANY` must be handled as well as Gilles mentions. @mswarowsky please open a CR on TF-M's gerrit and I will review it.

Reply:
I got other failtures when just easing the the check in cryptocell so have to debug further, that's why I went for the faster version of changing Mbed TLS. 
Will take that once I have time or do you have an issue tracker for TF-M to open it there.

Reply:
public issue tracker for TF-M is currently not working anymore - I will make an internal ticket to follow up on that as well.